SDWebImage/SDWebImageDownloaderOperation.m

```diff
@@ -317,48 +322,55 @@ - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)data
     if ((self.options & SDWebImageDownloaderProgressiveDownload) && self.expectedSize > 0) {
         // The following code is from http://www.cocoaintheshell.com/2011/05/progressive-images-download-imageio/
         // Thanks to the author @Nyx0uf
-
+        
+        // Get the image data
+        NSData *imageData = [self.imageData copy];
         // Get the total bytes downloaded
-        const NSInteger totalSize = self.imageData.length;
-
+        const NSInteger totalSize = imageData.length;
+        // Get the finish status
+        BOOL finished = (self.expectedSize == totalSize);
+        
+        if (!_imageSource) {
+            _imageSource = CGImageSourceCreateIncremental(NULL);
+        }
         // Update the data source, we must pass ALL the data, not just the new bytes
-        CGImageSourceRef imageSource = CGImageSourceCreateWithData((__bridge CFDataRef)self.imageData, NULL);
-
-        if (width + height == 0) {
-            CFDictionaryRef properties = CGImageSourceCopyPropertiesAtIndex(imageSource, 0, NULL);
+        CGImageSourceUpdateData(_imageSource, (__bridge CFDataRef)imageData, finished);
+        
+        if (_width + _height == 0) {
+            CFDictionaryRef properties = CGImageSourceCopyPropertiesAtIndex(_imageSource, 0, NULL);
             if (properties) {
                 NSInteger orientationValue = -1;
                 CFTypeRef val = CFDictionaryGetValue(properties, kCGImagePropertyPixelHeight);
-                if (val) CFNumberGetValue(val, kCFNumberLongType, &height);
+                if (val) CFNumberGetValue(val, kCFNumberLongType, &_height);
                 val = CFDictionaryGetValue(properties, kCGImagePropertyPixelWidth);
-                if (val) CFNumberGetValue(val, kCFNumberLongType, &width);
+                if (val) CFNumberGetValue(val, kCFNumberLongType, &_width);
                 val = CFDictionaryGetValue(properties, kCGImagePropertyOrientation);
                 if (val) CFNumberGetValue(val, kCFNumberNSIntegerType, &orientationValue);
                 CFRelease(properties);
-
+                
                 // When we draw to Core Graphics, we lose orientation information,
                 // which means the image below born of initWithCGIImage will be
                 // oriented incorrectly sometimes. (Unlike the image born of initWithData
                 // in didCompleteWithError.) So save it here and pass it on later.
 #if SD_UIKIT || SD_WATCH
-                orientation = [[self class] orientationFromPropertyValue:(orientationValue == -1 ? 1 : orientationValue)];
+                _orientation = [[self class] orientationFromPropertyValue:(orientationValue == -1 ? 1 : orientationValue)];
 #endif
             }
         }
-
-        if (width + height > 0 && totalSize < self.expectedSize) {
```

Comment:
This seems strange to when `totalSize > self.expectedSize`. But we can keep this to protect the accident condition.